<HTML>

<HEAD>
<TITLE>
BINF3020/BINF9020 - Assignment 1
</TITLE>
</HEAD>

<BODY>

<H2 ALIGN="center">
BINF3020/BINF9020 - Term 3, 2022
</H2>

<H3>
Assignment 1 - Protein sequence evolution and alignment
</H3>

<H4>
Allocation: assignment 1 is worth 20% of your final assessment.
</H4>

<P>
The aim of this assignment is for you to write software in C, Java, Python
(or alternative by arrangement) and run it on some biological sequence
data to generate outputs which you will write up in a comprehensive
report. The assignment will test your understanding of some key concepts
of evolutionary modelling and sequence alignment, and will require you to
demonstrate good written communication skills.

<P>
The assignment is structured to have three deliverable parts, as follows.

<DL>
<DT>
Part 1
<DD>
implement a protein sequence evolution simulator [worth up to 35% of total mark, i.e., 7/20]
<DT>
Part 2
<DD>
implement alignment by dynamic programming [worth up to 45% of total mark, i.e., 9/20]
<DT>
Part 3
<DD>
run your software to simulate the evolution of a protein sequence over
increasing evolutionary time, and align the mutated sequences with the
original. Write a report including a plot of percentage identity
against evolution time and discuss the results [worth up to 20% of total mark, i.e., 4/20]
</DL>

<P>
Each part is built upon the previous part and you should complete each
part before moving on to the next.

<P>
The marking will be based on having a running program for each part
and writing a proper report. Marks will also be given for good programming
and writing style. Marks will be deducted for program bugs and other errors.

<P>
The standard warnings against plagiarism apply and will definitely be enforced.
<BR>
Penalties for late submission of assignment parts are under consideration.

<HR>

<H4>
Part 2
</H4>

<br>
<b>NOTE:</b>
in this part of the assignment you have a choice between two versions of the algorithm to be implemented.

Description of the task:

<p>
You will implement pairwise global alignment of protein sequences in C, Java or
Python (or other language ONLY by prior agreement with course convenor).
The method will be the Needleman-Wunsch dynamic programming algorithm
using one of the two standard <em>gap cost</em> methods for scoring gaps
mentioned in lectures: either a <em>linear</em> or an <em>affine</em> gap
cost.

<p>
You should base your implementation on the BINF3020/BINF9020 lecture notes covering the
Needleman-Wunsch algorithm ("Global Alignment").

<p>
If you decide to implement Needleman-Wunsch with the <em>linear gap
cost</em> method for Part 2 the maximum mark available will be <b>7/20</b>
since this level of difficulty is about the same as Part 1 of the
assignment (the evolution simulator).

<p>
On the other hand, if you decide to implement Needleman-Wunsch with the
<em>affine gap cost</em> method for Part 2 the maximum mark available will
be <b>9/20</b> to reflect the slightly higher level of difficulty.

<p>
If you are in doubt about which to version of the algorithm to implement
it is <b>strongly recommended</b> to ensure you can get an implementation
of Needleman-Wunsch with the linear gap penalty working before attempting
a version using the affine gap penalty.

<p> 
Of course, if you are confident you can implement the affine gap penalty version
then just go ahead.

<p>
Ensure you use the following values to implement the gap penalty:
<br>
<b>Linear:</b> use a gap cost <i> d = 8 </i>
<br>
<b>Affine:</b> use a gap opening cost <i> d = 7 </i> and a gap extension cost <i> e = 1 </i>.

<p>
<b>These penalty values should be set using constants internally in your code!</b>

<p>
<b>File naming:</b> there will be two versions of the filename stem, depending on
whether your implementation is for (1) linear gap, or (2) affine gap
penalty. Of course, your filename should also have the correct language-specific extension
(<tt>.c</tt>, <tt>.py</tt>, <tt>.java</tt>, etc.).

<p>
For example, if you implemented Needleman-Wunsch with the linear gap
penalty in C, then your filename will be:
<br>
<tt> align1.c </tt>
<br>
However, if you implemented Needleman-Wunsch with the affine gap penalty
in Python, then your filename will be:
<br>
<tt> align2.py </tt>
<br>
Whatever language you choose for your implementation, you must indicate
which gap penalty method you have implemented by using the correct
filename stem (<tt>align1</tt> or <tt>align2</tt>) !

<p>
Your program needs to handle the following task.
The output file generated by your program for Part 1 contains, in order, your
original personal unique amino acid sequence followed by 500 sequences
representing its evolutionary divergence over 500 mutation cycles.
The alignment program should find the optimal alignment score of the original
sequence with itself and with each of the 500 mutated sequences. The output
for each alignment will contain two values. First will be the alignment score
returned by the dynamic programming algorithm, and second the <I>percentage</I>
of identities in the alignment.
Note that you <b>don't</b> have to generate the actual alignment, just these scores !

<P>
<B>
BLOSUM62 amino acid substitution matrix:
</B>

<TABLE>
<TR> <TD>  <TD> A <TD> B <TD> C <TD> D <TD> E <TD> F <TD> G <TD> H <TD> I <TD> K <TD> L <TD> M <TD> N <TD> P <TD> Q <TD> R <TD> S <TD> T <TD> V <TD> W <TD> X <TD> Y <TD> Z
<TR> <TD> A <TD> 4 <TD> -2 <TD> 0 <TD> -2 <TD> -1 <TD> -2 <TD> 0 <TD> -2 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -2 <TD> -1 <TD> -1 <TD> -1 <TD> 1 <TD> 0 <TD> 0 <TD> -3 <TD> -1 <TD> -2 <TD> -1
<TR> <TD> B <TD> -2 <TD> 6 <TD> -3 <TD> 6 <TD> 2 <TD> -3 <TD> -1 <TD> -1 <TD> -3 <TD> -1 <TD> -4 <TD> -3 <TD> 1 <TD> -1 <TD> 0 <TD> -2 <TD> 0 <TD> -1 <TD> -3 <TD> -4 <TD> -1 <TD> -3 <TD> 2
<TR> <TD> C <TD> 0 <TD> -3 <TD> 9 <TD> -3 <TD> -4 <TD> -2 <TD> -3 <TD> -3 <TD> -1 <TD> -3 <TD> -1 <TD> -1 <TD> -3 <TD> -3 <TD> -3 <TD> -3 <TD> -1 <TD> -1 <TD> -1 <TD> -2 <TD> -1 <TD> -2 <TD> -4
<TR> <TD> D <TD> -2 <TD> 6 <TD> -3 <TD> 6 <TD> 2 <TD> -3 <TD> -1 <TD> -1 <TD> -3 <TD> -1 <TD> -4 <TD> -3 <TD> 1 <TD> -1 <TD> 0 <TD> -2 <TD> 0 <TD> -1 <TD> -3 <TD> -4 <TD> -1 <TD> -3 <TD> 2
<TR> <TD> E <TD> -1 <TD> 2 <TD> -4 <TD> 2 <TD> 5 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> 1 <TD> -3 <TD> -2 <TD> 0 <TD> -1 <TD> 2 <TD> 0 <TD> 0 <TD> -1 <TD> -2 <TD> -3 <TD> -1 <TD> -2 <TD> 5
<TR> <TD> F <TD> -2 <TD> -3 <TD> -2 <TD> -3 <TD> -3 <TD> 6 <TD> -3 <TD> -1 <TD> 0 <TD> -3 <TD> 0 <TD> 0 <TD> -3 <TD> -4 <TD> -3 <TD> -3 <TD> -2 <TD> -2 <TD> -1 <TD> 1 <TD> -1 <TD> 3 <TD> -3
<TR> <TD> G <TD> 0 <TD> -1 <TD> -3 <TD> -1 <TD> -2 <TD> -3 <TD> 6 <TD> -2 <TD> -4 <TD> -2 <TD> -4 <TD> -3 <TD> 0 <TD> -2 <TD> -2 <TD> -2 <TD> 0 <TD> -2 <TD> -3 <TD> -2 <TD> -1 <TD> -3 <TD> -2
<TR> <TD> H <TD> -2 <TD> -1 <TD> -3 <TD> -1 <TD> 0 <TD> -1 <TD> -2 <TD> 8 <TD> -3 <TD> -1 <TD> -3 <TD> -2 <TD> 1 <TD> -2 <TD> 0 <TD> 0 <TD> -1 <TD> -2 <TD> -3 <TD> -2 <TD> -1 <TD> 2 <TD> 0
<TR> <TD> I <TD> -1 <TD> -3 <TD> -1 <TD> -3 <TD> -3 <TD> 0 <TD> -4 <TD> -3 <TD> 4 <TD> -3 <TD> 2 <TD> 1 <TD> -3 <TD> -3 <TD> -3 <TD> -3 <TD> -2 <TD> -1 <TD> 3 <TD> -3 <TD> -1 <TD> -1 <TD> -3
<TR> <TD> K <TD> -1 <TD> -1 <TD> -3 <TD> -1 <TD> 1 <TD> -3 <TD> -2 <TD> -1 <TD> -3 <TD> 5 <TD> -2 <TD> -1 <TD> 0 <TD> -1 <TD> 1 <TD> 2 <TD> 0 <TD> -1 <TD> -2 <TD> -3 <TD> -1 <TD> -2 <TD> 1
<TR> <TD> L <TD> -1 <TD> -4 <TD> -1 <TD> -4 <TD> -3 <TD> 0 <TD> -4 <TD> -3 <TD> 2 <TD> -2 <TD> 4 <TD> 2 <TD> -3 <TD> -3 <TD> -2 <TD> -2 <TD> -2 <TD> -1 <TD> 1 <TD> -2 <TD> -1 <TD> -1 <TD> -3
<TR> <TD> M <TD> -1 <TD> -3 <TD> -1 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> -2 <TD> 1 <TD> -1 <TD> 2 <TD> 5 <TD> -2 <TD> -2 <TD> 0 <TD> -1 <TD> -1 <TD> -1 <TD> 1 <TD> -1 <TD> -1 <TD> -1 <TD> -2
<TR> <TD> N <TD> -2 <TD> 1 <TD> -3 <TD> 1 <TD> 0 <TD> -3 <TD> 0 <TD> 1 <TD> -3 <TD> 0 <TD> -3 <TD> -2 <TD> 6 <TD> -2 <TD> 0 <TD> 0 <TD> 1 <TD> 0 <TD> -3 <TD> -4 <TD> -1 <TD> -2 <TD> 0
<TR> <TD> P <TD> -1 <TD> -1 <TD> -3 <TD> -1 <TD> -1 <TD> -4 <TD> -2 <TD> -2 <TD> -3 <TD> -1 <TD> -3 <TD> -2 <TD> -2 <TD> 7 <TD> -1 <TD> -2 <TD> -1 <TD> -1 <TD> -2 <TD> -4 <TD> -1 <TD> -3 <TD> -1
<TR> <TD> Q <TD> -1 <TD> 0 <TD> -3 <TD> 0 <TD> 2 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> 1 <TD> -2 <TD> 0 <TD> 0 <TD> -1 <TD> 5 <TD> 1 <TD> 0 <TD> -1 <TD> -2 <TD> -2 <TD> -1 <TD> -1 <TD> 2
<TR> <TD> R <TD> -1 <TD> -2 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> 2 <TD> -2 <TD> -1 <TD> 0 <TD> -2 <TD> 1 <TD> 5 <TD> -1 <TD> -1 <TD> -3 <TD> -3 <TD> -1 <TD> -2 <TD> 0
<TR> <TD> S <TD> 1 <TD> 0 <TD> -1 <TD> 0 <TD> 0 <TD> -2 <TD> 0 <TD> -1 <TD> -2 <TD> 0 <TD> -2 <TD> -1 <TD> 1 <TD> -1 <TD> 0 <TD> -1 <TD> 4 <TD> 1 <TD> -2 <TD> -3 <TD> -1 <TD> -2 <TD> 0
<TR> <TD> T <TD> 0 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -2 <TD> -2 <TD> -2 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> 0 <TD> -1 <TD> -1 <TD> -1 <TD> 1 <TD> 5 <TD> 0 <TD> -2 <TD> -1 <TD> -2 <TD> -1
<TR> <TD> V <TD> 0 <TD> -3 <TD> -1 <TD> -3 <TD> -2 <TD> -1 <TD> -3 <TD> -3 <TD> 3 <TD> -2 <TD> 1 <TD> 1 <TD> -3 <TD> -2 <TD> -2 <TD> -3 <TD> -2 <TD> 0 <TD> 4 <TD> -3 <TD> -1 <TD> -1 <TD> -2
<TR> <TD> W <TD> -3 <TD> -4 <TD> -2 <TD> -4 <TD> -3 <TD> 1 <TD> -2 <TD> -2 <TD> -3 <TD> -3 <TD> -2 <TD> -1 <TD> -4 <TD> -4 <TD> -2 <TD> -3 <TD> -3 <TD> -2 <TD> -3 <TD> 11 <TD> -1 <TD> 2 <TD> -3
<TR> <TD> X <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1 <TD> -1
<TR> <TD> Y <TD> -2 <TD> -3 <TD> -2 <TD> -3 <TD> -2 <TD> 3 <TD> -3 <TD> 2 <TD> -1 <TD> -2 <TD> -1 <TD> -1 <TD> -2 <TD> -3 <TD> -1 <TD> -2 <TD> -2 <TD> -2 <TD> -1 <TD> 2 <TD> -1 <TD> 7 <TD> -2
<TR> <TD> Z <TD> -1 <TD> 2 <TD> -4 <TD> 2 <TD> 5 <TD> -3 <TD> -2 <TD> 0 <TD> -3 <TD> 1 <TD> -3 <TD> -2 <TD> 0 <TD> -1 <TD> 2 <TD> 0 <TD> 0 <TD> -1 <TD> -2 <TD> -3 <TD> -1 <TD> -2 <TD> 5
</TABLE>

<p>
Be aware that this matrix contains rows and columns for three extra
amino acid code letters than were in your mutation matrix, namely
B, X and Z. This should not affect the alignment of your mutated sequences
but it should make your alignment program more general, i.e. able to
handle other sequences. <BR>
A version of this BLOSUM62 amino acid substitution matrix in comma-separated
format for copy/pasting into your source code file is at the end of this page. <BR>


<p>
The basic steps in your program will be to read in the initial sequence
and the 500 mutations of it from input. Then use your global alignment
procedures to do the following. First align the initial sequence <I>S0</I>
with itself, computing the score for the global alignment, and the percentage
of identities in the alignment.
In this case the length of the alignment is the same as the length of the
two identical sequences, and the percentage of identities will clearly be 100%.
Then for each of the remaining 500 mutated sequences <I>Si</I> in turn,
align <I>S0</I> with <I>Si</I>, computing the score for the global
alignment, and the percentage of identities in the alignment.
These alignment scores for every sequence from <I> 0 </I> to <I> 500 </I>
should be saved on a separate line to a single output file. 
The index number of the sequence should precede the scores on each line.

<P>
To compute the percentage of identities requires adding two counters, say
<I>k</I> and <I>id</I>, to the traceback part of global alignment algorithm,
as shown in the lecture notes. Before starting each alignment traceback
set <I>k</I> and <I>id</I> = 0. Then at each step in the traceback you
increment <I>k</I>, so <I>k</I> ends up being the length of the alignment.
Also, at each step in the traceback <U> where a residue <I>a</I> from sequence
<I>S0</I> is matched with a residue <I>b</I> from sequence <I>Si</I> </U>
the counter <I>id</I> is incremented by 1 if <I>a</I> is <B> IDENTICAL </B>
to <I>b</I>. The final percentage of identities in the alignment is
100 * (<I>id</I> / <I>k</I>).

<P>
Inputs and Outputs:
<TABLE>
<TR><TD> s501 <TD> the input file containing the initial sequence plus all
500 mutated sequences in FASTA format
<TR><TD> a501 <TD> the output file containing the self-alignment score and
percentage of identities for the initial sequence plus the alignment scores and
percentage of identities for all 500 mutated sequences
</TABLE>

<P>
In more detail, the output file should contain, on separate lines,
the sequence number and the alignment score and the percentage of identities
for each sequence.
The initial sequence is numbered zero, and its alignment score is for
self-alignment. Call this <I> v0 </I>. The percentage of identities in the
alignment of this sequence with itself is obviously 100%.
Call this <I> p0 </I>.
All mutated sequences are numbered from 1 to 500, and their alignment
scores are with respect to the initial sequence. For sequence <I> Si </I>
the score is <I> vi </I> and the percentage of identities is <I> pi </I>.
The order is the same as that in which they appear in the input file.
The output should form three columns, like this:
<TABLE>
<TR> <TH> N <TH> Score <TH> Percentage Identities
<TR> <TD> 0 <TD> <I> v0 </I> <TD> <I> p0 </I> 
<TR> <TD> 1 <TD> <I> v1 </I> <TD> <I> p1 </I>
<TR> <TD> 2 <TD> <I> v2 </I> <TD> <I> p2 </I>
<TR> <TD> 3 <TD> <I> v3 </I> <TD> <I> p3 </I>
<TR> <TD> ... <TD> ... <TD> ...
<TR> <TD> 500 <TD> <I> v500 </I> <TD> <I> p500 </I>
</TABLE>



<!--
% align < sequences > alignments
OR
% evolve < sequence | align > alignments
--> 
<p>
Your implementation should be designed to run as part of a command-line pipeline.

Therefore you should assume that the I/O comes from stdin and goes to stdout. <BR>
For example, the program would be executed as follows: <BR>

<p>
<tt>
$ alignN < s501 > a501
</tt>

<P>
where <tt>N</tt> is either 1 or 2, depending on the gap penalty method
implemented, <tt>s501</tt> is the file containing the evolved sequences
(plus the original) from your evolution simulator implemented in Part 1 of
the assignment, and <tt>a501</tt> is the corresponding output file in the
format described above.

<p>
Note that, if you are not implementing in a compiled language like C, the
invocation will look slightly different, depending on your setup.
For example, it may look like this: <br>

<p>
<tt> $ python3 ./alignN.py < s501 > a501
</tt>

<p>
However, <b>please ensure you include details on how to run your code</b> in a comment
at the beginning of your source file !

<P>
Additionally, you should be able to run both your programs as a pipeline
in a Unix shell, like this: <BR>
<P>
<TT>
$ evolve < s001.fasta | alignN > a501
</TT>
<P>
where <tt>s001.fasta</tt> is a valid FASTA format file, such as the original
sequence file assigned to you.

<P>
The only input routines you need are to read sequences in FASTA format.
If you made your routines modular for Part 1 you can re-use them in Part 2.
You will need to write out to a file containing, on each line, the sequence
number and the alignment score and number of identities.

<P>
Your program should report an error and halt if the input is not in FASTA
format or the sequence contains something other than the original 20 amino acid
code letters.


<P>
<B>
Deadline:
</B>
<BR>
You should submit Part 2 <U>by 23:59:59 on Sunday October  9, 2022.</U><BR>

<P>
<b>
Submission:
</b>
See Moodle course page for submission.


<P>
Your submission will be a single source file "alignN.c", "alignN.java", or
"alignN.py" (or agreed alternative) where "N" is either 1 or 2 as outlined
above, which should be adequately commented. <BR>

<br>
<b>NOTE:</b> 
<!-- be sure to test that your program works correctly on CSE machines running Linux <i>before</i> submitting it - you may lose marks if it does not work correctly on a CSE machine running Linux, e.g., one of the lab machines. -->
if possible, please ensure that your program works correctly on CSE machines running Linux <i>before</i> submitting it.
However, if this is <i>not</i> possible, for example, because your code uses an allowed module not installed on CSE machines, please include in your program details on the version of the imported module and
your software development environment (Windows, Linux or Mac OS versions, etc.).


<P>
A version of the BLOSUM62 amino acid substitution matrix in comma-separated
format for copy/pasting into your source file: <BR>

<P>
,A,B,C,D,E,F,G,H,I,K,L,M,N,P,Q,R,S,T,V,W,X,Y,Z <BR>
A,4,-2,0,-2,-1,-2,0,-2,-1,-1,-1,-1,-2,-1,-1,-1,1,0,0,-3,-1,-2,-1 <BR>
B,-2,6,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-1,-3,2 <BR>
C,0,-3,9,-3,-4,-2,-3,-3,-1,-3,-1,-1,-3,-3,-3,-3,-1,-1,-1,-2,-1,-2,-4 <BR>
D,-2,6,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-1,-3,2 <BR>
E,-1,2,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-1,-2,5 <BR>
F,-2,-3,-2,-3,-3,6,-3,-1,0,-3,0,0,-3,-4,-3,-3,-2,-2,-1,1,-1,3,-3 <BR>
G,0,-1,-3,-1,-2,-3,6,-2,-4,-2,-4,-3,0,-2,-2,-2,0,-2,-3,-2,-1,-3,-2 <BR>
H,-2,-1,-3,-1,0,-1,-2,8,-3,-1,-3,-2,1,-2,0,0,-1,-2,-3,-2,-1,2,0 <BR>
I,-1,-3,-1,-3,-3,0,-4,-3,4,-3,2,1,-3,-3,-3,-3,-2,-1,3,-3,-1,-1,-3 <BR>
K,-1,-1,-3,-1,1,-3,-2,-1,-3,5,-2,-1,0,-1,1,2,0,-1,-2,-3,-1,-2,1 <BR>
L,-1,-4,-1,-4,-3,0,-4,-3,2,-2,4,2,-3,-3,-2,-2,-2,-1,1,-2,-1,-1,-3 <BR>
M,-1,-3,-1,-3,-2,0,-3,-2,1,-1,2,5,-2,-2,0,-1,-1,-1,1,-1,-1,-1,-2 <BR>
N,-2,1,-3,1,0,-3,0,1,-3,0,-3,-2,6,-2,0,0,1,0,-3,-4,-1,-2,0 <BR>
P,-1,-1,-3,-1,-1,-4,-2,-2,-3,-1,-3,-2,-2,7,-1,-2,-1,-1,-2,-4,-1,-3,-1 <BR>
Q,-1,0,-3,0,2,-3,-2,0,-3,1,-2,0,0,-1,5,1,0,-1,-2,-2,-1,-1,2 <BR>
R,-1,-2,-3,-2,0,-3,-2,0,-3,2,-2,-1,0,-2,1,5,-1,-1,-3,-3,-1,-2,0 <BR>
S,1,0,-1,0,0,-2,0,-1,-2,0,-2,-1,1,-1,0,-1,4,1,-2,-3,-1,-2,0 <BR>
T,0,-1,-1,-1,-1,-2,-2,-2,-1,-1,-1,-1,0,-1,-1,-1,1,5,0,-2,-1,-2,-1 <BR>
V,0,-3,-1,-3,-2,-1,-3,-3,3,-2,1,1,-3,-2,-2,-3,-2,0,4,-3,-1,-1,-2 <BR>
W,-3,-4,-2,-4,-3,1,-2,-2,-3,-3,-2,-1,-4,-4,-2,-3,-3,-2,-3,11,-1,2,-3 <BR>
X,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1 <BR>
Y,-2,-3,-2,-3,-2,3,-3,2,-1,-2,-1,-1,-2,-3,-1,-2,-2,-2,-1,2,-1,7,-2 <BR>
Z,-1,2,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-1,-2,5 <BR>


<HR>

<H4>
Part 3
</H4>

To be announced
<!----------------------------------------------------------------------------->

<table align=right>
<tr><td><font size=-1>Last modified
<!-- Mon 26 Sep 2022 17:13:41 AEST -->
Mon 26 Sep 2022 19:10:35 AEST
</font></td></tr>
</table>


</BODY>
</HTML>
